### PR TITLE
[CDS-76320]: Update run-a-terraform-plan-with-the-terraform-plan-step.md

### PR DESCRIPTION
**Pull request by tmacari-harness**
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?
[CDS-76320]: Update run-a-terraform-plan-with-the-terraform-plan-step.md
- [ ] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ ] Tested Locally
- [ ] *Optional* Screen Shoot. 

[CDS-76320]: https://harness.atlassian.net/browse/CDS-76320?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/continuous-delivery/cd-infrastructure/terraform-infra/rollback-provisioned-infra-with-the-terraform-rollback-step.md [hunk 1]
```diff
@@ -48,6 +48,10 @@ Let's say you deployed two modules successfully already: module1 and module2. Ne
 
 However, let's look at the situation where module3 succeeds and now you have module1, module2, and module3 deployed. If the next deployment fails, the rollback will only roll back to the Terraform state with module3 deployed. Module1 and module2 weren't in the previous Terraform state, so the rollback excludes them.
 
+Rollback is not possible if you run apply with **Skip state storage** option and no terraform backed configured in your terraform files.
+Having Rollback step in such scenario would be a wrong setup and may end-up with an unexpected result.
```
Comment (michaelcretzman) [inline docs/continuous-delivery/cd-infrastructure/terraform-infra/rollback-provisioned-infra-with-the-terraform-rollback-step.md:52]:
```suggestion
Rollback is not possible if you run the Terraform Apply step with the **Skip state storage** option enabled and no Terraform backed is configured in your Terraform files.
Using the Rollback step in such a scenario would be an incorrect setup and might cause an unexpected result.
```

File: docs/continuous-delivery/cd-infrastructure/terraform-infra/run-a-terraform-plan-with-the-terraform-apply-step.md [hunk 2]
```diff
@@ -527,6 +527,21 @@ Here's an example of decrypted Terraform JSON output:
...
+That state file will be downloaded in terraform working directory for next executions then updated state will be uploaded back after execution ends.
+
+This way we allow keeping state of the infrastructure even if there is no terraform backend configured.
+This to is more for testing purpose, for productin environments we advise to cofigure a backend in your terarform config files, ref: [https://developer.hashicorp.com/terraform/language/settings/backends/configuration](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).
+
+With **Skip state storage** option we allow you to skip local state upload and download operations mentioned above.
+This option makes sense only if you do not have terarform backed configured in your terraform config files, if you have terarform backed configured then terraform cli will not create any local state files.
+
+If terarform apply step is configured to skip state storage and there is no backend configured in your terraform files, you should not add a rollback step as this is an invalid setup.
+Rollback is impossible if there is no state file.
```
Comment (michaelcretzman) [inline docs/continuous-delivery/cd-infrastructure/terraform-infra/run-a-terraform-plan-with-the-terraform-apply-step.md:543]:
```suggestion
## Skip state storage

While running Terraform commands on the delegate, Harness by default will try so detect if there is a local state file in the Terraform working directory.

If local state file is identified, at the end of the execution it is saved on Harness storage with a key based on the provisioner identifier.

That state file is downloaded in the Terraform working directory for subsequent executions, and the updated state is uploaded after execution ends.

This method allows the maintaining of the state of the infrastructure even if there is no Terraform backend configured.

This is more for testing purposes. For production environments, Harness advises you configure a backend in your Terraform config files. For information, go to [Backend Configuration](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).

With the **Skip state storage** option enabled, Harness allows you to skip the local state upload and download operations mentioned above.

This option makes is useful only if you do not have a Terraform backed configured in your Terraform config files. If you have a Terraform backed configured, then the Terraform CLI will not create any local state files.

If the Terraform Apply step is configured to skip state storage, and there is no backend configured in your Terraform files, you should not add a rollback step, as this is an invalid setup. Rollback is impossible if there is no state file.
```

File: docs/continuous-delivery/cd-infrastructure/terraform-infra/run-a-terraform-plan-with-the-terraform-plan-step.md [hunk 3]
```diff
@@ -554,6 +554,19 @@ Terraform has compared your real infrastructure against your configuration
...
+
+While running terraform commands on delegate, Harness bay defaul will try so detect if there is a local state file in terraform working directory.
+If local state file is identified, at the end of the execution it will be saved on harness storage with a key created based on provisioner identifier.
+That state file will be downloaded in terraform working directory for next executions then updated state will be uploaded back after execution ends.
+
+This way we allow keeping state of the infrastructure even if there is no terraform backend configured.
+This to is more for testing purpose, for productin environments we advise to cofigure a backend in your terarform config files, ref: [https://developer.hashicorp.com/terraform/language/settings/backends/configuration](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).
+
+With **Skip state storage** option we allow you to skip local state upload and download operations mentioned above.
+This option makes sense only if you do not have terarform backed configured in your terraform config files, if you have terarform backed configured then terraform cli will not create any local state files.
```
Comment (michaelcretzman) [inline docs/continuous-delivery/cd-infrastructure/terraform-infra/run-a-terraform-plan-with-the-terraform-plan-step.md:567]:
```suggestion
While running Terraform commands on the delegate, Harness by default will try so detect if there is a local state file in the Terraform working directory.

If local state file is identified, at the end of the execution it is saved on Harness storage with a key based on the provisioner identifier.

That state file is downloaded in the Terraform working directory for subsequent executions, and the updated state is uploaded after execution ends.

This method allows the maintaining of the state of the infrastructure even if there is no Terraform backend configured.

This is more for testing purposes. For production environments, Harness advises you configure a backend in your Terraform config files. For information, go to [Backend Configuration](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).

With the **Skip state storage** option enabled, Harness allows you to skip the local state upload and download operations mentioned above.

This option makes is useful only if you do not have a Terraform backed configured in your Terraform config files. If you have a Terraform backed configured, then the Terraform CLI will not create any local state files.

```

**Review by michaelcretzman — DISMISSED** [review]

File: docs/continuous-delivery/cd-infrastructure/terraform-infra/rollback-provisioned-infra-with-the-terraform-rollback-step.md [hunk 4]
```diff
@@ -34,7 +34,7 @@ terraform --version
 
 ## Review: Terraform Rollback
 
-When rollback happens, Harness rolls back the provisioned infrastructure to the previous successful version of the Terraform state.
+When rollback happens, Harness rolls back the provisioned infrastructure to the previous successful version of the Terraform state and using config files/terraform configuration from latest successful deployment with matching  **Provisioner Identifier**.
```
Comment (michaelcretzman) [inline docs/continuous-delivery/cd-infrastructure/terraform-infra/rollback-provisioned-infra-with-the-terraform-rollback-step.md:37]:
```suggestion
When rollback is performed, Harness rolls back the provisioned infrastructure to the previous successful version of the Terraform state using config files or the Terraform configuration from the latest successful deployment with a matching  **Provisioner Identifier**.
```

**Comment by tmacari-harness** [comment]
Looks good for me.
Thank you Michael.

**Pull request merged**